### ISSUE TITLE
feat: pattern option

### ISSUE THREAD

**Issue by Eomm** [LyraSearch/plugin-docusaurus#77]
If the user can set the `pattern` option:

https://github.com/LyraSearch/plugin-docusaurus/blob/846ee5eb69373f22ddadd09d907250fc68d01205/src/server/index.ts#L33

It should be able to archive this use case:

The fastify website will have ~15 documentation versions that lead to a >10MB Lyra's index that is not usable (slow, bandwidth consuming).

So, by filtering the pattern, it should be able for the user to trim all the docs' versions that must not be indexed. For fastify, it would index only 1 version for each major:

Example usage:

```
plugins: [
    [
      '@lyrasearch/plugin-docusaurus',
      {
        pattern: "(version-latest|version-3.x|version-2.x)/**/*.html"
      },
    ],
  ],
```

**Comment by micheleriva** [comment]
I was facing the same problem with Lyra's docs. I like this proposed solution!

**Comment by Ceres6** [comment]
I think you are describing different approaches, one is filtering which versions to index at build time (in `docusaurus.config.js`) and the other is having indexes for each and downloading/using just the index of the version the user is currently viewing. I like the second better. WDYT?

**Comment by Ceres6** [comment]
Yes, I believe you can get all versions and create an index for each one using the glob pattern, then you just need to modify the `retreiveIndex` function to get the version you're currently at

**Comment by stearm** [comment]
@Ceres6 isn't the `retrieveIndex` called only at build time? Instead, the docs version the user is reading is defined at runtime. `useFetchIndex` should be modified, is there something I'm missing?

**Comment by Ceres6** [comment]
@stearm Yes, sorry, I meant the `getLyraSearch` function

**Comment by Eomm** [comment]
My 2 cent: do I want to index all the `html` pages?

I would provide the pattern option in any case to the user

**Comment by Ceres6** [comment]
We could implement that feature, although I'm not sure if that should be a priority right now if we choose the second approach

**Comment by Eomm** [comment]
Yeah, sure! I don't suggest any priority or dead line 🙌🏽

**Comment by stearm** [comment]
I'm wondering if Docusaurus has a rule for putting the docs in the build directory, I can't find anything inside the documentation about the build folder structure. I'm comparing two builds: fastify docs and lyra.

**Fastify**
Here we have all the docs inside the `/docs` folder, so it is quite easy to create the indexes.

![fastify](https://user-images.githubusercontent.com/14802795/212038699-6468fa8d-8760-446c-a782-a6b862d749e9.png)

**Lyra**
Here we have the current one on the root, without a parent and the others (v0.3.1) on the same level.

![lyra](https://user-images.githubusercontent.com/14802795/212038726-72bd7334-eaa8-451c-a598-6bd799793f90.png)

Do you have any suggestion? Am I missing something?

**Comment by Eomm** [comment]
When a dev uses Docusaurus has two main dev-folders:

- `/docs`: where the canary/next release is written
- `/versioned_docs`: where all the older document versions will be moved

So, the current fastify website-next does not rely on the `/docs` folder - it has only `/versioned_docs`. (since the legacy website has the `docs/latest` URL it was the only way I found to support such URLs)

Instead, lyra is using both directories.

I think these docs will be helpful:

- https://docusaurus.io/docs/versioning
- https://docusaurus.io/docs/versioning#terminology

**Comment by stearm** [comment]
> So, the current fastify website-next does not rely on the `/docs` folder

Sorry, maybe I did not explain myself. I'm speaking only about the `build` (or `lib`) folder, in which you can find compiled html, that the plugin uses to create indexes.

The glob pattern is referencing to the build folder, not to the src. Isn't it?

https://github.com/LyraSearch/plugin-docusaurus/blob/846ee5eb69373f22ddadd09d907250fc68d01205/src/server/index.ts#L33

**Comment by Ceres6** [comment]
I think it is actually referencing the parent one so it will find both of them

**Comment by stearm** [comment]
> > 'm speaking only about the build (or lib) folder
> 
> What I meant is that those dev folders impact the `./build` output:
> 
> * `./docs/` is being transpiled to `./build`
> * `./versioned_docs` folder's content is being moved to `./build/<version_n>`

Thank you for clarifying. We were able to write a solution that followed this pattern. I'll keep you informed.

**Comment by stearm** [comment]
@Eomm I created #79 to fetch only the index the user needs, then we can add a `pattern` option.

**Comment by stearm** [comment]
The only thing we need to work on is managing custom version names (paths different from the version name), therefore I don't have much time to work on it . I still have a commit to push anyway. Please get in touch with me if you need anything and feel free to keep working or get ideas by this PR.

**Comment by Ceres6** [comment]
I don't think anyone is

**Comment by micheleriva** [comment]
This would be a good candidate for hub's work @simoneb
